packages/@aws-cdk/cdk/lib/core/tag-manager.ts

```diff
@@ -0,0 +1,149 @@
+ * the `Construct` creates using `toToken()` for lazy evaluations
+ */
+export class TagManager {
+
+    /**
+     * Checks if the object implements the `ITaggable` interface
+     */
+    public static isTaggable(taggable: ITaggable | any): taggable is ITaggable {
+        return ((taggable as ITaggable).tags !== undefined);
+    }
+
+    private readonly _tags: ITags = {};
+
+    private readonly parentTagManager: TagManager | undefined;
+
+    constructor(partner: Construct, initialTags: ITags = {}) {
+        if (TagManager.isTaggable(partner.parent)) {
+            this.parentTagManager = partner.parent.tags;
+        } else {
+            this.parentTagManager = undefined;
+        }
+        this._tags = this.defaultPropagate(initialTags);
+    }
+
+    /**
+     * Returns a deep copy of the tags object, except for nested Tokens
+     *
+     * Deep copies of all primiatve `ITagProps` are returned. If a `value` is a
+     * `Token` then that reference is kept and not deep copied. The assumption is
+     * this is for lazy evaluation. Modifying a nested `Token` is not advised
+     * the consumer should use `addTags` with the corresponding key to change
+     * the value or override a propagated parent tag.
+     */
+    public get tags(): ITags {
+        const tags: ITags = {};
+        Object.keys(this._tags).map( key => (tags[key] = {...this._tags[key]}));
+        return {...tags, ...this.propagatedTags()};
+    }
+
+    /**
+     * Converts the `tags` to a Token for use in lazy evaluation
+     */
+    public toToken(propagateOnly = false): Token {
+        return new Token(() => this.toArray(propagateOnly));
+    }
+
+    /**
+     * Returns a deep copy of the tags as `Array<{key: string, value: any}>`
+     */
+    public toArray(propagateOnly = false): ITag[] {
+        if (propagateOnly) {
+            const propTags = this.propagatedTags();
+            return Object.keys(propTags).map( key => ({key, value: propTags[key].value}));
+        }
+        const tags = this.tags;
+        return Object.keys(tags).map( key => ({key, value: tags[key].value}));
+    }
+
+    /**
+     * Adds the specified tag to the array of tags
+     *
+     * @param tag The tag to add
+     * @param propagate  Whether to add this tag to all children
+     */
+    public addTags(tags: ITags): void {
+        Object.assign(this._tags, this.defaultPropagate(tags));
+    }
+
+    /**
+     * Removes the specified tag from the array if it exists
+     *
+     * @param key The key of the tag to remove
+     * @param propagate  Whether to remove this tag from all children
+     */
+    public removeTag(key: string): void {
+        delete this._tags[key];
+    }
+
+    /**
+     * Check that a key exists
+     */
+    public hasTag(key: string): boolean {
+        return this._tags[key] !== undefined;
+    }
+
+    /**
+     * Retrieve all propagated tags including tags from parents
+     */
+    private propagatedTags(): ITags {
```

Comment:
Since parent tags are mutable, this won't really work... Can this be lazy? Instead of fetching all tags from parent upon init, can we do this only during synthesis (i.e. `toToken`).

Reply:
This is used in the `toCloudformation()` call and is lazy there? I am not sure if you are implying I should always wrap this in a `Token` or if I should add warnings to the consumer in the doc string?